frost-secp256k1-tr/src/lib.rs

```diff
@@ -327,68 +304,176 @@ impl Ciphersuite for Secp256K1Sha256 {
+            Cow::Borrowed(signer_nonces),
+            Cow::Owned(key_package.clone().into_even_y(None)),
+        ))
+    }
+
+    // Preprocess sign inputs, negating the keys in the PublicKeyPackage if
+    // required by BIP-340.
+    fn pre_aggregate<'a>(
+        _ctx: &mut Self::Context,
+        signing_package: &'a SigningPackage,
+        signature_shares: &'a BTreeMap<Identifier, round2::SignatureShare>,
+        public_key_package: &'a keys::PublicKeyPackage,
+    ) -> Result<
+        (
+            Cow<'a, SigningPackage>,
+            Cow<'a, BTreeMap<Identifier, round2::SignatureShare>>,
+            Cow<'a, keys::PublicKeyPackage>,
+        ),
+        Error,
+    > {
+        Ok((
+            Cow::Borrowed(signing_package),
+            Cow::Borrowed(signature_shares),
+            Cow::Owned(public_key_package.clone().into_even_y(None)),
+        ))
+    }
+
+    // Preprocess verify inputs, negating the VerifyingKey if required by
+    // BIP-340.
+    fn pre_verify<'a>(
+        message: &'a [u8],
+        signature: &'a Signature,
+        public_key: &'a VerifyingKey,
+    ) -> Result<(Cow<'a, [u8]>, Cow<'a, Signature>, Cow<'a, VerifyingKey>), Error> {
+        let public_key = public_key.into_even_y(None);
+        Ok((
+            Cow::Borrowed(message),
+            Cow::Borrowed(signature),
+            Cow::Owned(public_key),
+        ))
+    }
+
+    // Generate a nonce, negating it if required by BIP-340.
+    fn generate_nonce<R: RngCore + CryptoRng>(
+        rng: &mut R,
+    ) -> (
+        <<Self::Group as Group>::Field as Field>::Scalar,
+        <Self::Group as Group>::Element,
+    ) {
+        let k = random_nonzero::<Self, R>(rng);
+        let R = <Self::Group>::generator() * k;
+        if R.to_affine().y_is_odd().into() {
+            (-k, -R)
+        } else {
+            (k, R)
+        }
+    }
+
+    // Compute the group commitment, negating if required by BIP-340. Note that
+    // only at this point it is possible to check if negation will be required
+    // or not. If it is, it will require negating the participant's nonces (when
+    // signing) or their group commitment share (when aggregating). This is
+    // signaled by setting the `is_group_commitment_even` flag in the Context.
+    fn compute_group_commitment(
+        ctx: &mut Self::Context,
+        signing_package: &SigningPackage,
+        binding_factor_list: &frost_core::BindingFactorList<S>,
+    ) -> Result<GroupCommitment<S>, Error> {
+        let group_commitment = compute_group_commitment(signing_package, binding_factor_list)?;
+        ctx.is_group_commitment_even =
+            (!group_commitment.clone().to_element().to_affine().y_is_odd()).into();
+        let group_commitment = if !ctx.is_group_commitment_even {
+            GroupCommitment::<S>::from_element(-group_commitment.to_element())
+        } else {
+            group_commitment
+        };
+
+        Ok(group_commitment)
+    }
```

Comment:
This sounds great! I'd like to avoid having to pass the group commitment around if we could, but this feels much better than having to add `Context`. Please go ahead if you want, thanks.

Reply:
@conradoplg I've pushed these changes, `Context` is no more! :rocket: